File: bundles/org.eclipse.vjet.vsf.vjo.kernel/src/org/eclipse/vjet/vjo/util/JsonVjo.java
/*******************************************************************************
 * Copyright (c) 2012 eBay Inc. and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     eBay Inc. - initial API and implementation
 *******************************************************************************/
package org.eclipse.vjet.vjo.util;

import org.eclipse.vjet.dsf.javatojs.anno.AJsProxy;
import org.eclipse.vjet.dsf.dap.proxy.NativeJsProxy;
import org.mozilla.mod.javascript.Scriptable;
import org.eclipse.vjet.dsf.javatojs.anno.AExclude;
import org.eclipse.vjet.dsf.dap.proxy.INativeJsFuncProxy;
import org.eclipse.vjet.dsf.dap.proxy.NativeJsFuncProxy;
import org.eclipse.vjet.dsf.dap.proxy.NativeJsTypeRef;
import org.eclipse.vjet.dsf.javatojs.anno.AJavaOnly;

//NativeJsProxy for vjo.util.JsonVjo.js
@org.eclipse.vjet.dsf.resource.utils.CodeGen("NativeJsProxyGenerator")
@AJsProxy
public class JsonVjo extends NativeJsProxy {

    /** for framework use only */
    @AExclude
    public JsonVjo(Scriptable nativeObj){
        super(nativeObj);
    }

    /** internal use only */
    protected JsonVjo(Object ...args){
        super(args);
    }

    public JsonVjo() {
        super();
    }

    public static Object json2vjo(String jsonStr, String typeName) {
        return callStaticWithName("vjo.util.JsonVjo", "json2vjo", Object.class, jsonStr, typeName);
    }

    @AJavaOnly
    public static final NativeJsTypeRef<JsonVjo> prototype = NativeJsTypeRef.get(JsonVjo.class);

    public static final INativeJsFuncProxy<NativeJsTypeRef<JsonVjo>> json2vjo = NativeJsFuncProxy.create(prototype, "json2vjo");
}

File: vsf/bundles/org.eclipse.vjet.vsf.vjo.kernel/bin/org/eclipse/vjet/vjo/util/JsonVjo.js
/*******************************************************************************
 * Copyright (c) 2012 eBay Inc. and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     eBay Inc. - initial API and implementation
 *******************************************************************************/
vjo.ctype('org.eclipse.vjet.vjo.util.JsonVjo')
.props({
	//> public Object json2vjo(String jsonStr, String typeName)
	json2vjo : function (jsonStr, typeName) {
		// need to trim the type from JS string, otherwise it will confuse the JS deserialization code
		var i = jsonStr.indexOf(':');
		jsonStr = jsonStr.substring(i+1, jsonStr.length-1);
		
		// call toVjo() to convert json string to VJO object
		return this.toVjo(typeName, eval("(" + jsonStr + ")"));
	},
	//> private void json2vjo(String type, Object json)
	toVjo : function toVjo (type, json) {
		if (typeof(type) == 'undefined' || type == null || type === '') {
			throw "Invalid type name";
		}
		if (typeof(json) == 'string' && type == "float") {
		   return parseFloat(json);
		}
		if (typeof(json) == 'string' && type == "number") {
		   return parseInt(json);
		}
		if (!json) return null;
	
		var TypeObj, err = "Invalid type '" + type + "'.";
		try {
			TypeObj = eval(type);
		} catch (e) {
			throw err;
		}
		if (!TypeObj) return err;
	
		if (TypeObj.clazz._type == "etype") {
			var o = "enum";
		} else {
			var o = new TypeObj();
			for (var i in json) {
				var outElementName = i;
				if (TypeObj.clazz.elementNameMap != null) {
				   if (TypeObj.clazz.elementNameMap[i] != null) {
					outElementName = TypeObj.clazz.elementNameMap[i];
				   }
				}
	
				// find the type of the field by traversing the meta map of itself and its parents
				var fieldType = vjo.findType(TypeObj, outElementName);
				if (fieldType != null) {
					var idx = fieldType.indexOf(':');
					if (idx != -1) {
						// ":" is found. this is an array
						var arrayClassName = fieldType.substring(0, idx);
						var ArrayType = eval(arrayClassName);
						var arrayObj = new ArrayType();
						var elementType = fieldType.substring(idx+1, fieldType.length);
						var array = json[i];
	
						if (elementType == "String" || elementType == "int" || elementType == "long") {
						   // iterate thru the list and add one by one
						   for (var j in array) {
							arrayObj.add(array[j]);
						   }
						} else {
						   // even the meta says it's an array, it could just be a single element,
						   // in which case it will be flattened.
						   if (vjo.isArray(array)) {
						      // need to recurse into each element for deserialization
						      for (var j in array) {
							var nestedObj = toVjo(elementType, array[j]);
							arrayObj.add(nestedObj);
						      }
						   } else {
						      var nestedObj  = toVjo(elementType, array);
						      arrayObj.add(nestedObj);
						   }
						}
						// at the end, set arrryObj
						o[outElementName] = arrayObj;
					} else {
						var nestedObj = toVjo(fieldType, json[i]);
						o[outElementName] = nestedObj;
					}
				} else {
					o[outElementName] = json[i];
				}
			}
		}
	
		return o;
	}
})
.endType();